libmisc/salt.c

```diff
@@ -263,6 +267,78 @@ static /*@observer@*/const char *gensalt_bcrypt (void)
 }
 #endif /* USE_BCRYPT */
 
+#ifdef USE_YESCRYPT
+/* Default cost if not explicitly specified.  */
+#define Y_COST_DEFAULT 8
+/* Minimum cost.  */
+#define Y_COST_MIN 1
+/* Maximum cost.  */
+#define Y_COST_MAX 11
+/*
+ * Return a salt prefix specifying the cost for the YESCRYPT method.
+ */
+static /*@observer@*/const char *YESCRYPT_salt_cost (/*@null@*/long *prefered_cost)
+{
+	static char cost_prefix[5];
+        long cost;
+
+        if (NULL == prefered_cost) {
+	    cost = getdef_num ("YESCRYPT_COST_FACTOR", Y_COST_DEFAULT);
+        } else {
+            cost = *prefered_cost;
+        }
+
+        if (cost < Y_COST_MIN) {
+            cost = Y_COST_MIN;
+        }
+        if (cost > Y_COST_MAX) {
+            cost = Y_COST_MAX;
+        }
+
+        cost_prefix[0] = 'j';
+        if (cost < 3) {
+            cost_prefix[1] = 0x36 + cost;
+        } else if (cost < 6) {
+            cost_prefix[1] = 0x34 + cost;
+        } else {
+            cost_prefix[1] = 0x3b + cost;
+        }
+        cost_prefix[2] = cost >= 3 ? 'T' : '5';
+        cost_prefix[3] = '$';
+        cost_prefix[4] = 0;
+
+        return cost_prefix;
+}
+
+/*
+ * Default number of base64 characters used for the salt.
+ * 24 characters gives a 144 bits (18 bytes) salt. Unlike the more
+ * traditional 128 bits (16 bytes) salt, this 144 bits salt is always
+ * represented by the same number of base64 characters without padding
+ * issue, even with a non-standard base64 encoding scheme.
+ */
+#define YESCRYPT_SALT_SIZE 24
+/*
+ *  Generate a 22 character salt string for yescrypt.
+ */
+static /*@observer@*/const char *gensalt_yescrypt (void)
+{
+	static char salt[32];
+
+	salt[0] = '\0';
+
+	seedRNG ();
+	strcat (salt, l64a (random()));
```

Comment:
This leading `strcat()` call looks unneeded as the loop contains the same (will just do one more iteration of this line is omitted).

Reply:
I copy/pasted this code from [`gensalt_bcrypt()`](https://github.com/shadow-maint/shadow/blob/c917ed7b76c31a9aa100477c6c4d281c7d835c93/libmisc/salt.c#L248-L264), which seems to be an adaptation from [`gensalt()`](https://github.com/shadow-maint/shadow/blob/c917ed7b76c31a9aa100477c6c4d281c7d835c93/libmisc/salt.c#L272-L289). Since I do not fully understand how this code works, I didn't touched it except for the salt's length.

Reply:
Makes sense to reuse the same (non-ideal) code for this PR - in fact, this is why I didn't complain about the uses of `random()`. I should have realized the redundant `strcat()` probably came from there as well.

I wouldn't increase salt size. Sure 144 bits gets encoded more optimally, but it's excessive and besides with `random()` we only have at most 32 bits for real: https://github.com/shadow-maint/shadow/blob/e84df9e163e133eb11a2728024ff3e3440592cf8/libmisc/salt.c#L72

The Owl patch I referred to above also switched to use of `/dev/urandom`, but of course this shouldn't be in this same PR.

BTW, it's weird the same code was used to generate salts for bcrypt, since bcrypt uses the crypt base64 alphabet in different order than all other crypts. This probably ended up relying on the bcrypt implementation truncating the unsupported bit values for the last character, probably leaving slightly fewer than 128 bits.

Reply:
> I wouldn't increase salt size. Sure 144 bits gets encoded more optimally

Oh, maybe the reason you did that is the salt strings truncated at 22 were not being accepted, because they contained extra 4 bits in the last char, and you didn't feel like implementing encoding of exactly 128 bits?

Reply:
It's not really that I don't "feel like" implementing it. My reasoning was that 144 bits does not harm in any way (libxcrypt's `man 5 crypt` states that the salt can be up to 512 bits) and truncating to 128 bits, while not adding value, would require me to take time study the undocumented non-standard base64 encoding with the risk of introducing bugs in my implementation. Less code, less time spent, less risks and no loss in terms of security, all the odds were in favor of a 144 bits salt.

Obviously, if you recommend to use 128 bits anyway (I'm not sure what you imply by saying it's excessive), I can do that.

Reply:
> the undocumented non-standard base64 encoding

For salt and hash, yescrypt uses the exact same standard crypt base64 encoding that traditional Unix crypt(3) did since the 1970s. It's also the encoding that md5crypt, sha256crypt, and sha512crypt use. (bcrypt is an outlier here.)

However, the yescrypt implementation is strict in that it only accepts salts that are exactly the same as what would be in the resulting salt+hash string. This is different from bcrypt, which silently turns the last character of salt into how it actually sees it in binary, re-encoding it in the returned salt+hash string. This is also different from md5crypt, sha256crypt, and sha512crypt, which use the salt as a mostly opaque ASCII string, preserving the full 132 bits. So it's probably because of these differences that the same code producing 132-bit salts didn't work for yescrypt here, while it did for others. Not the different base64 encoding.

Sure 144 bits does not harm, it's just different. By "excessive", I was referring to 144 vs. 128, not to 128.